### PR TITLE
Select2 Upgrade for sms settings Page

### PR DESCRIPTION
**Pull request by Rohit25negi**
Upgrade select2 from v3 to v4 on `/a/<domain>/sms/settings/`

### CODE REVIEW

**Comment by Rohit25negi** [comment]
This can be reviewed. But not to merge before https://github.com/dimagi/commcare-hq/pull/22025

File: corehq/apps/hqwebapp/static/hqwebapp/js/select2_handler_v4.js [hunk 1]
```diff
@@ -16,6 +16,7 @@ hqDefine("hqwebapp/js/select2_handler_v4", [
         var self = {};
         self.fieldName = options.fieldName;
         self.placeholder = options.placeholder;
+        self.allowClear = options.allowClear;
```
Comment (emord) [inline corehq/apps/hqwebapp/static/hqwebapp/js/select2_handler_v4.js]: Is this change related to the SMS changes? I don't see where it's being used. Also should it be defaulted to `true` to keep the same behavior?

Comment (Rohit25negi) [reply corehq/apps/hqwebapp/static/hqwebapp/js/select2_handler_v4.js]: Yes, On sms setting page option clearing is not allowed.
I also thought of making default to true. But looking that this(v4) is only used only 1 another place where it was not very reasonable to have  default `true` therefore I did not make it default. 
Rethought on this you are right i make sense to make it default else other new places using(v4) have to explicitly set it `true`.
Making it default

**Review by biyeun — APPROVED** [review]

File: corehq/apps/hqwebapp/static/hqwebapp/js/select2_handler_v4.js [hunk 2]
```diff
@@ -16,6 +16,7 @@ hqDefine("hqwebapp/js/select2_handler_v4", [
         var self = {};
         self.fieldName = options.fieldName;
         self.placeholder = options.placeholder;
+        self.allowClear = options.allowClear || true;
```
Comment (emord) [inline corehq/apps/hqwebapp/static/hqwebapp/js/select2_handler_v4.js]: this will always be true. I think that you want something like:

```
self.allowClear = options.allowClear !== undefined ? options.allowClear : true
```

Comment (Rohit25negi) [reply corehq/apps/hqwebapp/static/hqwebapp/js/select2_handler_v4.js]: Oops!. Sorry for this. slipped from mind in hurry. Next time will be more careful. 
Done correcting

**Review by emord — APPROVED** [review]

**Review by orangejenny — CHANGES_REQUESTED** [review]
When I check this out locally and go to http://localhost:8000/a/DOMAIN/sms/settings/, the dropdown for Default Case Owner throws a 500: `ValueError: Cannot use None as a query value` - any idea what's going on there?

Also, did you test this on staging? I think it will break in a production-like environment, because now `sms/js/settings` is using v4, but langcodes, which uses v3, is also part of the SMS bundle. You can view the bundle config by running `build_requirejs` in your branch (after refreshing the `staticfiles` dir by deleting it and running `collectstatic`) and then viewing `staticfiles/build.txt` - if a single bundle contains both `select2-3.5.2-legacy/select2.min.js` and `select2/dist/js/select2.full.min.js` it's a problem.

Because of this, select2 migrations generally need to be done on an entire app, not on individual pages. The dependencies get messy and fragile. Because langcodes is also used in app manager, which is still on select2 v3, langcodes will likely need to be split into a v3 and v4.

**Comment by Rohit25negi** [comment]
@orangejenny  500 error which is coming is because of bug which solved here https://github.com/dimagi/commcare-hq/pull/22025. Since it is merged now, it will work after merge with that(Thought will be anyway merged after pr review therefore did not explicitly merged this from master).

About Langcode problem. It seems like pretty dependent code. will look into it to solve it.

**Comment by orangejenny** [comment]
> 500 error which is coming is because of bug which solved here #22025

Cool, can you merge master into this PR? That 500 is making it impossible to test that the settings page can be saved, and this migration can break saving.

> It seems like pretty dependent code.

Indeed, this select2 migration is fragile. I'm hoping to get the full migration finished soon (this quarter) so we can just forget about it.
